modules/renter/contractor/negotiate.go

```diff
@@ -71,48 +78,116 @@ func negotiateContract(conn net.Conn, host modules.HostDBEntry, fc types.FileCon
+	if err := encoding.ReadObject(conn, &response, modules.MaxErrorSize); err != nil {
 		return Contract{}, errors.New("couldn't read the host's response to our proposed contract: " + err.Error())
 	}
 	if response != modules.AcceptResponse {
 		return Contract{}, errors.New("host rejected proposed contract: " + response)
 	}
-	var hostTxnSet []types.Transaction
-	if err := encoding.ReadObject(conn, &hostTxnSet, types.BlockSizeLimit); err != nil {
-		return Contract{}, errors.New("couldn't read the host's updated contract: " + err.Error())
+	// host now sends any new parent transactions, inputs and outputs that
+	// were added to the transaction
+	var newParents []types.Transaction
+	var newInputs []types.SiacoinInput
+	var newOutputs []types.SiacoinOutput
+	if err := encoding.ReadObject(conn, &newParents, types.BlockSizeLimit); err != nil {
+		return Contract{}, errors.New("couldn't read the host's added parents: " + err.Error())
+	}
+	if err := encoding.ReadObject(conn, &newInputs, types.BlockSizeLimit); err != nil {
+		return Contract{}, errors.New("couldn't read the host's added inputs: " + err.Error())
+	}
+	if err := encoding.ReadObject(conn, &newOutputs, types.BlockSizeLimit); err != nil {
+		return Contract{}, errors.New("couldn't read the host's added outputs: " + err.Error())
 	}
 
-	// TODO: check txn?
+	// merge txnAdditions with txnSet
+	txnBuilder.AddParents(newParents)
+	for _, input := range newInputs {
+		txnBuilder.AddSiacoinInput(input)
+	}
+	for _, output := range newOutputs {
+		txnBuilder.AddSiacoinOutput(output)
+	}
+
+	// sign the txn
+	signedTxnSet, err := txnBuilder.Sign(true)
+	if err != nil {
+		return Contract{}, err
+	}
+
+	// calculate signatures added by the transaction builder
+	var addedSignatures []types.TransactionSignature
+	_, _, _, addedSignatureIndices := txnBuilder.ViewAdded()
+	for _, i := range addedSignatureIndices {
+		addedSignatures = append(addedSignatures, signedTxnSet[len(signedTxnSet)-1].TransactionSignatures[i])
+	}
+
+	// Send acceptance and signatures
+	// TODO: validate new txn
+	if err := encoding.WriteObject(conn, modules.AcceptResponse); err != nil {
+		return Contract{}, errors.New("couldn't send transaction acceptance: " + err.Error())
+	}
+	if err := encoding.WriteObject(conn, addedSignatures); err != nil {
+		return Contract{}, errors.New("couldn't send added signatures: " + err.Error())
+	}
+
+	// Read the host signatures.
+	var acceptStr string
+	err = encoding.ReadObject(conn, &acceptStr, modules.MaxErrorSize)
+	if err != nil {
+		return Contract{}, err
+	}
+	if acceptStr != modules.AcceptResponse {
+		return Contract{}, errors.New(acceptStr)
+	}
+	var hostSigs []types.TransactionSignature
+	if err := encoding.ReadObject(conn, &hostSigs, 2e3); err != nil {
+		return Contract{}, errors.New("couldn't read the host's signatures: " + err.Error())
+	}
+	for _, sig := range hostSigs {
+		txnBuilder.AddTransactionSignature(sig)
+	}
```

Comment:
I'm not going to spend a ton of time reviewing the networking code for now, I think we're still tweaking things. Once it feels like the networking code has hit a more or less resting state (including all RPCs implemented), then I'll look at it more closely.

Also, I do think that we should try to get the renew rpc working this release. I said earlier that we didn't need it, but we're in the RPC headspace so we should probably knock them all out at once.